core/bridges/cache.go

```diff
@@ -0,0 +1,260 @@
+	err := c.ORM.DeleteBridgeType(ctx, bt)
+	if err != nil {
+		if !errors.Is(err, sql.ErrNoRows) {
+			return err
+		}
+	}
+
+	// We delete regardless of the rows affected, in case it gets out of sync
+	c.bridgeTypesCache.Delete(bt.Name)
+
+	return err
+}
+
+func (c *Cache) BridgeTypes(ctx context.Context, offset int, limit int) ([]BridgeType, int, error) {
+	return c.ORM.BridgeTypes(ctx, offset, limit)
+}
+
+func (c *Cache) CreateBridgeType(ctx context.Context, bt *BridgeType) error {
+	err := c.ORM.CreateBridgeType(ctx, bt)
+	if err != nil {
+		return err
+	}
+
+	c.bridgeTypesCache.Store(bt.Name, *bt)
+
+	return nil
+}
+
+func (c *Cache) UpdateBridgeType(ctx context.Context, bt *BridgeType, btr *BridgeTypeRequest) error {
+	if err := c.ORM.UpdateBridgeType(ctx, bt, btr); err != nil {
+		return err
+	}
+
+	c.bridgeTypesCache.Store(bt.Name, *bt)
+
+	return nil
+}
+
+func (c *Cache) GetCachedResponse(ctx context.Context, dotId string, specId int32, maxElapsed time.Duration) ([]byte, error) {
+	// prefer to get latest value from cache
+	c.mu.RLock()
+	cached, inCache := c.bridgeLastValueCache[responseKey(dotId, specId)]
+	c.mu.RUnlock()
+
+	if inCache && cached.FinishedAt.After(time.Now().Add(-maxElapsed)) {
+		return cached.Value, nil
+	}
+
+	response, finishedAt, err := c.ORM.GetCachedResponseWithFinished(ctx, dotId, specId, maxElapsed)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mu.Lock()
+	c.bridgeLastValueCache[responseKey(dotId, specId)] = BridgeResponse{
+		DotID:      dotId,
+		SpecID:     specId,
+		Value:      response,
+		FinishedAt: finishedAt,
+	}
+	c.mu.Unlock()
+
+	return response, nil
+}
+
+func (c *Cache) UpsertBridgeResponse(ctx context.Context, dotId string, specId int32, response []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.bridgeLastValueCache[responseKey(dotId, specId)] = BridgeResponse{
+		DotID:      dotId,
+		SpecID:     specId,
+		Value:      response,
+		FinishedAt: time.Now(),
+	}
+
+	return nil
+}
+
+func (c *Cache) Start(context.Context) error {
+	return c.StartOnce(CacheServiceName, func() error {
+		c.wg.Add(1)
+
+		go c.run()
+
+		return nil
+	})
+}
+
+func (c *Cache) Close() error {
+	return c.StopOnce(CacheServiceName, func() error {
+		c.stop <- struct{}{}
+		c.wg.Wait()
+
+		return nil
+	})
+}
+
+func (c *Cache) Ready() error {
+	return c.StateMachine.Ready()
+}
+
+func (c *Cache) HealthReport() map[string]error {
+	return map[string]error{CacheServiceName: c.Healthy()}
+}
+
+func (c *Cache) Name() string {
+	return CacheServiceName
+}
+
+func (c *Cache) run() {
+	defer c.wg.Done()
+
+	for {
+		timer := time.NewTimer(utils.WithJitter(c.interval))
+
+		select {
+		case <-timer.C:
+			c.doBulkUpsert()
+		case <-c.stop:
+			timer.Stop()
+
+			return
+		}
+	}
+}
+
+func (c *Cache) doBulkUpsert() {
+	c.mu.RLock()
+	values := make([]BridgeResponse, 0, len(c.bridgeLastValueCache))
+
+	for _, value := range c.bridgeLastValueCache {
+		values = append(values, value)
+	}
+
```

Comment:
```suggestion
	values := slices.Clone(c.bridgeLastValueCache)
```

Reply:
This is a slice from a map, not a new slice from an existing slice. Is there a shortcut method for this?

Reply:
Ah, yeah there is: https://pkg.go.dev/golang.org/x/exp/maps#Values